lib/ExecutionEngine/ExecutionEngine.cpp

```diff
@@ -266,3 +266,100 @@ void ExecutionEngine::compile(Function *F, CompilationContext &cctx,
     EXIT_ON_ERR(funcOrErr.takeError());
   }
 }
+
+namespace {
+/// The name of the temporary function to be used to perform constant folding.
+constexpr const char *constEvaluationFunctionName =
+    "__constEvaluationFunction__";
+
+/// \returns true of a node is a constant operation, i.e. it is a constant or
+/// all of its inputs are constants.
+bool isConstantOperation(const Node *N) {
+  if (llvm::isa<Constant>(N) || llvm::isa<SplatNode>(N)) {
+    return true;
+  }
+  if (llvm::isa<Placeholder>(N)) {
+    return false;
+  }
+  for (unsigned idx = 0, e = N->getNumInputs(); idx < e; ++idx) {
+    auto input = N->getNthInput(idx);
+    if (!llvm::isa<Constant>(input) && !llvm::isa<SplatNode>(input)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+/// Evaluates a provided constant operation \p C in the context of the function
+/// \p F.
+/// \returns a constant representing a result of computation.
+Node *evaluateConstantOperation(Function *F, Node *C) {
+  assert(isConstantOperation(C) && "Expected a constant expression");
+  // Create a temporary execution engine to compute only this specific
+  // operation.
+  PlaceholderBindings bindings;
+  ExecutionEngine EE("Interpreter");
+  Module &mod = *F->getParent();
+  // Create a temporary function to perform the constant operation.
+  Function *constEvaluationF = mod.createFunction(constEvaluationFunctionName);
+  unsigned constIdx = 0;
+  // Mapping from existing nodes to the new ones.
+  NodeMap currToNew;
+  // Clone the constant operation and some of its inputs if necessary.
+  auto *clonedC = recursiveClone(constEvaluationF, C, currToNew);
+  SaveNode *SN = constEvaluationF->createSave("save", clonedC);
+  auto *outputPH = SN->getPlaceholder();
+  bindings.allocate(outputPH);
+  // Run the temporary execution engine to perform this constant operation
+  // evaluation.
+  EE.compile(CompilationMode::Infer, constEvaluationF);
+  EE.run(bindings);
+  // Get the result of the constant operation and create a new constant from it.
+  Tensor *outputTensor = bindings.get(outputPH);
+  // Remove the temporary function.
+  mod.eraseFunction(constEvaluationF);
+  // Return the newly created constant, which is the result of a constant
+  // operation.
+  return mod.createConstant(strFormat("const_%u", constIdx++), *outputTensor);
+}
+} // namespace
+
+bool glow::constantFold(Function *F, CompilationContext &cctx) {
+  // Avoid recursively constant folding a function that is submitted for
+  // constant folding through a temporary execution engine.
+  if (F->getName().startswith(constEvaluationFunctionName)) {
```

Comment:
I see why this is needed but slightly fragile. I wonder if there is another way to avoid this. I can't think of one right now, but maybe we can add ability to disable certain optimization passes in CompilationContext or maybe we need a more simple way to run the interpreter without doing optimizations (would also probably solve the circular dep issue)

Reply:
IMO we should/could just skip graph optimizations in this case. There's some weirdness here where the optimizer already also does Constant folding, e.g. our current `convertConstant()` optimizations that already do some of this Constant folding for `Quantize(Constant)` etc. In this case we may end up copying the results to a new buffer unnecessarily. This could be as easy as calling `backend->compile()` instead of `EE.compile()`. Not sure this helps with the circular dependency issue though.

Reply:
Yes, this is exactly what we discussed with @jfix71 last week. The reason I haven't done it yet is that is seemed to be not so trivial as I expected ;-)

Reply:
@jackm321  I tried having a compile light with optimizations disabled, but some of the ops actually depend on the lowering/optimization passes and tests started failing. One such example is reduceMean2AvgPool.

Reply:
Yeah this is true, we may need to at least lower.